### PR TITLE
[#5421] Enforce ILL in unlink/rename/trim (master)

### PR DESCRIPTION
**Pull request by alanking**
This passed CI tests except for 1 test which passes at the bench... and the failure doesn't seem... wrong, necessarily...

`irods.test.test_resource_types.Test_Resource_ReplicationToTwoCompoundResourcesWithPreferArchive.test_iget_with_purgec`

```
...

 --- IrodsSession: icommand executed by [otherrods] [iput purgecgetfile.txt] --- 
Assert Command: iput purgecgetfile.txt
Expecting EMPTY: ['']
  stdout:
    | 
  stderr:
    | 
Output found

 --- IrodsSession: icommand executed by [otherrods] [iget -f --purgec purgecgetfile.txt] --- 
Assert Command: iget -f --purgec purgecgetfile.txt
Expecting EMPTY: ['']
  stdout:
    | 
  stderr:
    | 
Output found

 --- IrodsSession: icommand executed by [otherrods] [ils -L purgecgetfile.txt] --- 
Assert FAIL Command: ils -L purgecgetfile.txt
Expecting STDOUT_SINGLELINE: [' 0 ', 'purgecgetfile.txt']
  stdout:
    |   otherrods         0 demoResc;compResc2;cacheResc2           54 2021-05-03.20:40 & purgecgetfile.txt
    |         generic    /var/lib/irods/cacheResc2Vault/home/otherrods/2021-05-04Z00:40:06--irods-testing-Tyx6Eb/purgecgetfile.txt
    |   otherrods         1 demoResc;compResc2;archiveResc2           54 2021-05-03.20:40 & purgecgetfile.txt
    |         generic    /var/lib/irods/archiveResc2Vault/home/otherrods/2021-05-04Z00:40:06--irods-testing-Tyx6Eb/purgecgetfile.txt
    |   otherrods         3 demoResc;compResc1;archiveResc1           54 2021-05-03.20:40 & purgecgetfile.txt
    |         generic    /var/lib/irods/archiveResc1Vault/home/otherrods/2021-05-04Z00:40:06--irods-testing-Tyx6Eb/purgecgetfile.txt
  stderr:
    | 
Output found

FAILED TESTING ASSERTION
...
```

It wants replica 0 to be gone, but I believe it just purged the cache copy on the hierarchy which was chosen for the get operation (demoResc;compResc1). So the test should be modified to look for the cache copy in the hierarchy that won the vote, right?

### CODE REVIEW

**Comment by trel** [comment]
i think that's right - it's not ALL cache copies... or a fixed cached copy... it's the one related to the get.

File: server/api/src/rsDataObjRename.cpp [hunk 1]
```diff
@@ -129,20 +132,46 @@ namespace
...
                     return status;
                 }
+
+                if (const auto ret = ill::try_lock(*dataObjInfoHead, ill::lock_type::write); ret < 0) {
+                    const auto msg = fmt::format(
+                        "rename not allowed because source data object is locked "
+                        "[error code=[{}], source path=[{}]",
+                        ret, srcDataObjInp->objPath);
+
+                    //addRErrorMsg(&rsComm->rError, ret, msg.data());
```
Comment (trel) [inline server/api/src/rsDataObjRename.cpp:150]: dead?

File: server/api/src/rsDataObjRename.cpp [hunk 2]
```diff
@@ -129,20 +132,46 @@ namespace
...
+                    // The force flag ensures that unlink will not send the object to the trash.
+                    destination_cond_input[FORCE_FLAG_KW] = "";
+
+                    if (const auto ec = rsDataObjUnlink(rsComm, destDataObjInp); ec < 0) {
+                        const auto msg = fmt::format(
+                            "failed to unlink destination data object for overwrite via rename "
+                            "[error code=[{}], source path=[{}], destination path=[{}]]",
+                            ec, srcDataObjInp->objPath, destDataObjInp->objPath);
+
+                        //addRErrorMsg(&rsComm->rError, ret, msg.data());
```
Comment (trel) [inline server/api/src/rsDataObjRename.cpp:168]: dead?

File: server/api/src/rsDataObjUnlink.cpp [hunk 3]
```diff
@@ -178,6 +180,13 @@ int rsMvDataObjToTrash(
         return status;
     }
 
+    if (const auto ret = ill::try_lock(**dataObjInfoHead, ill::lock_type::write); ret < 0) {
+        irods::log(LOG_NOTICE, fmt::format(
+            "[{}:{}] - move to trash not allowed because data object is locked",
```
Comment (trel) [inline server/api/src/rsDataObjUnlink.cpp:185]: do we want logical path here, too?

File: server/api/src/rsDataObjUnlink.cpp [hunk 4]
```diff
@@ -291,11 +300,27 @@ int rsMvDataObjToTrash(
 
             auto replica = *id::find_replica(op, std::stoi(cond_input.at(REPL_NUM_KW).value().data()));
 
+            if (const auto ret = ill::try_lock(*op.get(), ill::lock_type::write, replica.replica_number()); ret < 0) {
+                irods::log(LOG_NOTICE, fmt::format(
+                    "[{}:{}] - unlink not allowed because data object is locked",
```
Comment (trel) [inline server/api/src/rsDataObjUnlink.cpp:305]: do we want logical path / replica number here, too?

File: server/api/src/rsDataObjUnlink.cpp [hunk 5]
```diff
@@ -291,11 +300,27 @@ int rsMvDataObjToTrash(
...
+            }
+
             return dataObjUnlinkS(&_comm, &_inp, replica.get());
         }
 
         auto op = id::make_data_object_proxy(**_head);
 
+        if (const auto ret = ill::try_lock(*op.get(), ill::lock_type::write); ret < 0) {
+            irods::log(LOG_NOTICE, fmt::format(
+                "[{}:{}] - unlink not allowed because data object is locked",
```
Comment (trel) [inline server/api/src/rsDataObjUnlink.cpp:318]: do we want logical path here, too?

File: server/api/src/rsDataObjUnlink.cpp [hunk 6]
```diff
@@ -356,11 +381,9 @@ int rsMvDataObjToTrash(
                 addKeyVal(&dataObjUnlinkInp->condInput, RESC_HIER_STR_KW, hier.c_str());
             }
             catch (const irods::exception& e) {
-                std::stringstream msg;
-                msg << "failed in irods::resolve_resource_hierarchy for [";
-                msg << dataObjUnlinkInp->objPath << "]";
                 irods::log(PASS(irods::error(e)));
-                // TODO: If the force flag is set, we just continue?
+                // TODO: If the force flag is set, we just continue???
+                // TODO: See irods/irods#3154 and talk to JMC
```
Comment (trel) [inline server/api/src/rsDataObjUnlink.cpp:386]: maybe... just the issue number is sufficient :)

File: unit_tests/src/test_logical_locking.cpp [hunk 7]
```diff
@@ -51,18 +76,141 @@ TEST_CASE("basic write lock")
...
+    // Use a separate connection to run the tests from that used in the setup
+    irods::experimental::client_connection new_conn;
+    RcComm& new_comm = static_cast<RcComm&>(new_conn);
+
+    // Prepare input to rcDataObjOpen for the test cases
+    DataObjInp fail_open_inp{};
+    auto fail_cond_input = irods::experimental::make_key_value_proxy(fail_open_inp.condInput);
+    std::snprintf(fail_open_inp.objPath, sizeof(fail_open_inp.objPath), "%s", target_object.c_str());
+
+    //SECTION("open for read")
```
Comment (trel) [inline unit_tests/src/test_logical_locking.cpp:154]: this is commented out - not sure if that's intended.

File: unit_tests/src/test_logical_locking.cpp [hunk 8]
```diff
@@ -51,18 +76,141 @@ TEST_CASE("basic write lock")
...
+        // open existing, currently opened replica
+        fail_cond_input[RESC_HIER_STR_KW] = opened_replica_resc;
+        REQUIRE(INTERMEDIATE_REPLICA_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+
+        // open existing, currently locked replica
+        fail_cond_input[RESC_HIER_STR_KW] = other_replica_resc;
+        REQUIRE(LOCKED_DATA_OBJECT_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+    }
+
+    //SECTION("open for write")
```
Comment (trel) [inline unit_tests/src/test_logical_locking.cpp:167]: this is commented out - not sure if that's intended.

File: unit_tests/src/test_logical_locking.cpp [hunk 9]
```diff
@@ -51,18 +76,141 @@ TEST_CASE("basic write lock")
...
+        // open existing, currently opened replica
+        fail_cond_input[RESC_HIER_STR_KW] = opened_replica_resc;
+        REQUIRE(INTERMEDIATE_REPLICA_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+
+        // open existing, currently locked replica
+        fail_cond_input[RESC_HIER_STR_KW] = other_replica_resc;
+        REQUIRE(LOCKED_DATA_OBJECT_ACCESS == rcDataObjOpen(&new_comm, &fail_open_inp));
+    }
+
+    //SECTION("open for create")
```
Comment (trel) [inline unit_tests/src/test_logical_locking.cpp:180]: this is commented out - not sure if that's intended.

File: server/api/src/rsDataObjRename.cpp (hunk 1 above)
Comment (alanking) [reply server/api/src/rsDataObjRename.cpp:150]: I couldn't decide if there should be a message back to the user or just let the error code do the talking. There's also the matter of `imv -r` spewing a lot of messages, etc.

File: server/api/src/rsDataObjRename.cpp (hunk 2 above)
Comment (alanking) [reply server/api/src/rsDataObjRename.cpp:168]: See above

File: server/api/src/rsDataObjUnlink.cpp (hunk 3 above)
Comment (alanking) [reply server/api/src/rsDataObjUnlink.cpp:185]: Sure

File: server/api/src/rsDataObjUnlink.cpp (hunk 4 above)
Comment (alanking) [reply server/api/src/rsDataObjUnlink.cpp:305]: Done

File: server/api/src/rsDataObjUnlink.cpp (hunk 5 above)
Comment (alanking) [reply server/api/src/rsDataObjUnlink.cpp:318]: Done

File: server/api/src/rsDataObjUnlink.cpp (hunk 6 above)
Comment (alanking) [reply server/api/src/rsDataObjUnlink.cpp:386]: lol Forgot I wrote this. Done

File: unit_tests/src/test_logical_locking.cpp (hunk 7 above)
Comment (alanking) [reply unit_tests/src/test_logical_locking.cpp:154]: Oh, yeah, these sections cause the whole test case to reset each time so I removed the labels to save a little time but I don't know that it really buys us anything. I think adding them back makes sense.

File: unit_tests/src/test_logical_locking.cpp (hunk 8 above)
Comment (alanking) [reply unit_tests/src/test_logical_locking.cpp:167]: Added back the sections

File: unit_tests/src/test_logical_locking.cpp (hunk 9 above)
Comment (alanking) [reply unit_tests/src/test_logical_locking.cpp:180]: Added back the sections.

**Comment by alanking** [comment]
I opened #5593 for the matter of the failing test.

File: server/api/src/rsDataObjRename.cpp (hunk 1 above)
Comment (trel) [reply server/api/src/rsDataObjRename.cpp:150]: right, certainly don't want recursive being noisy.    let's leave it out for now - and let the error code do the work for now.   can always make it more helpful/friendly later.

File: server/api/src/rsDataObjRename.cpp (hunk 2 above)
Comment (trel) [reply server/api/src/rsDataObjRename.cpp:168]: torch

File: server/api/src/rsDataObjRename.cpp (hunk 1 above)
Comment (alanking) [reply server/api/src/rsDataObjRename.cpp:150]: Yep, I'll take it out, then

File: server/api/src/rsDataObjRename.cpp (hunk 2 above)
Comment (alanking) [reply server/api/src/rsDataObjRename.cpp]: Done

**Comment by alanking** [comment]
Force-pushed because rebased. I'll run it through CI one more time

**Comment by alanking** [comment]
CI tests passed

**Comment by trel** [comment]
also closed #4236

**Pull request merged**